### PR TITLE
Adopt Sendable in the Task APIs

### PR DESCRIPTION
**Pull request by DougGregor**
Rework `Sendable` checking to be completely based on "missing"
conformances, so that we can individually diagnose missing `Sendable`
conformances based on both the module in which the conformance check
happened as well as where the type was declared. The basic rules here
are to only diagnose if the non-`Sendable` type either comes from the
current model or from a module that was compiled in a mode that
consistently diagnoses `Sendable`. The latter is true for modules
compiled in Swift 6 mode or that provide `-warn-concurrency` on
the command line. This is tracked by rdar://78269348.

The diagnostics will be an error in Swift 6 or a warning in Swift 5.x.

Use this to adopt `Sendable` in the Task APIs without breaking code, 
tracked by rdar://77441933.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Review by ktoso — APPROVED** [review]

**Comment by DougGregor** [comment]
Welp, looks like I need to fix that compile error.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: test/Concurrency/concurrent_value_inference.swift [hunk 1]
```diff
@@ -10,7 +10,7 @@ struct S1 {
   var c: C2
 }
 
-enum E1 {
+enum E1 { // expected-note{{add conformance of enum 'E1' to the 'Sendable' protocol}}
```
Comment (karwa) [inline test/Concurrency/concurrent_value_inference.swift:13]: This message is a bit strange. We don’t add conformance _of_ types _to_ protocols. Might I suggest:

> add conformance to the ‘Sendable’ protocol to the enum ‘E1’.

or even

> add ‘Sendable’ conformance to the enum ‘E1’.

Comment (DougGregor) [reply test/Concurrency/concurrent_value_inference.swift:13]: We use the "consider" wording elsewhere, so I'm going to try out

```
consider making enum 'E1' conform to the 'Sendable' protocol 
```

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Review by ghost — APPROVED** [review]

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Pull request merged**